utils/frame/remote-externalities/src/lib.rs

```diff
@@ -320,72 +350,143 @@ where
 		Ok(keys)
 	}
 
-	/// Synonym of `rpc_get_pairs_unsafe` that uses paged queries to first get the keys, and then
+	/// Synonym of `getPairs` that uses paged queries to first get the keys, and then
 	/// map them to values one by one.
 	///
 	/// This can work with public nodes. But, expect it to be darn slow.
 	pub(crate) async fn rpc_get_pairs_paged(
 		&self,
 		prefix: StorageKey,
 		at: B::Hash,
+		pending_ext: &mut TestExternalities,
 	) -> Result<Vec<KeyValue>, &'static str> {
 		let keys = self.rpc_get_keys_paged(prefix, at).await?;
-		let keys_count = keys.len();
-		log::debug!(target: LOG_TARGET, "Querying a total of {} keys", keys.len());
+		let client = self.as_online().transport.remote_client.clone().unwrap();
+		let thread_chunk_size =
+			((keys.len() + self.as_online().threads - 1) / self.as_online().threads).max(1);
 
-		let mut key_values: Vec<KeyValue> = vec![];
-		let client = self.as_online().rpc_client();
-		for chunk_keys in keys.chunks(BATCH_SIZE) {
-			let batch = chunk_keys
-				.iter()
-				.cloned()
-				.map(|key| ("state_getStorage", rpc_params![key, at]))
-				.collect::<Vec<_>>();
-
-			let values = client.batch_request::<Option<StorageData>>(batch).await.map_err(|e| {
-				log::error!(
-					target: LOG_TARGET,
-					"failed to execute batch: {:?}. Error: {:?}",
-					chunk_keys.iter().map(HexDisplay::from).collect::<Vec<_>>(),
-					e
-				);
-				"batch failed."
-			})?;
+		log::info!(
+			target: LOG_TARGET,
+			"Querying a total of {} keys, splitting among {} threads, {} keys per thread",
+			keys.len(),
+			self.as_online().threads,
+			thread_chunk_size,
+		);
 
-			assert_eq!(chunk_keys.len(), values.len());
+		let mut handles = Vec::new();
+		let keys_chunked: Vec<Vec<StorageKey>> =
+			keys.chunks(thread_chunk_size).map(|s| s.into()).collect::<Vec<_>>();
 
-			for (idx, key) in chunk_keys.iter().enumerate() {
-				let maybe_value = values[idx].clone();
-				let value = maybe_value.unwrap_or_else(|| {
-					log::warn!(target: LOG_TARGET, "key {:?} had none corresponding value.", &key);
-					StorageData(vec![])
-				});
-				key_values.push((key.clone(), value));
-				if key_values.len() % (10 * BATCH_SIZE) == 0 {
-					let ratio: f64 = key_values.len() as f64 / keys_count as f64;
-					log::debug!(
-						target: LOG_TARGET,
-						"progress = {:.2} [{} / {}]",
-						ratio,
-						key_values.len(),
-						keys_count,
-					);
+		enum Message {
+			Terminated,
+			Batch(Vec<(Vec<u8>, Vec<u8>)>),
+		}
+		let (tx, rx) = std::sync::mpsc::channel::<Message>();
+
+		for thread_keys in keys_chunked {
+			let thread_client = client.clone();
+			let thread_sender = tx.clone();
+			let handle = std::thread::spawn(move || {
```

Comment:
I think these spawned threads doesn't really help make thing faster.

As the RPC client has a shared background thread which does the heavy work such spawning threads just to await for some async I/O to become ready seems like a waste.

I guess we could try with `FuturesUnordered/Ordered` here instead.

Reply:
The only benefit I can think of is that the batch requests can performed in parallel but for "waiting" for the responses it should help.

I still think awaiting on the responses is bottleneck here...

Reply:
Yeah my experiments also show that we have not gained much benefit from this. If confirmed, I will revert this part.

Reply:
Although, the nice thing about threading is that we can write to `pending_ext` little by little, and while downloading is happening, which is know to bring a good amount of improvements.